File: java/src/config/TomlConfig.java
package config;

import java.io.File;
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;

/**
 * Pure-Java TOML config parser matching the PowerShell/Rust TomlConfig.
 * Supports single-line key=value pairs and inline arrays.
 */
class TomlConfig {

    private String[] keys;
    private String[] values;
    private int size;
    private String[] profiles;
    private int profileCount;

    TomlConfig() {
        keys = new String[256];
        values = new String[256];
        size = 0;
        profiles = new String[32];
        profileCount = 0;
    }

    static TomlConfig load(String path) {
        var file = new File(path);
        if (!file.exists()) {
            System.err.println("Config file not found: " + path);
            return null;
        }

        var config = new TomlConfig();
        var currentSection = "";

        try {
            var reader = new BufferedReader(new FileReader(file));
            String rawLine;
            while ((rawLine = reader.readLine()) != null) {
                var line = stripComment(rawLine).trim();
                if (line.isEmpty()) continue;

                // Section header
                if (line.startsWith("[") && line.endsWith("]")) {
                    currentSection = line.substring(1, line.length() - 1);
                    if (currentSection.startsWith("profiles.")) {
                        var profileKey = currentSection.substring("profiles.".length());
                        config.profiles[config.profileCount] = profileKey;
                        config.profileCount++;
                    }
                    continue;
                }

                // Key = value
                int eq = line.indexOf('=');
                if (eq > 0) {
                    var key = line.substring(0, eq).trim();
                    var value = line.substring(eq + 1).trim();
                    var fullKey = currentSection.isEmpty() ? key : currentSection + "." + key;
                    config.keys[config.size] = fullKey;
                    config.values[config.size] = value;
                    config.size++;
                }
            }
            reader.close();
        } catch (IOException e) {
            System.err.println("Error reading config: " + e.getMessage());
            return null;
        }

        return config;
    }

    String getValue(String key) {
        for (int i = 0; i < size; i++) {
            if (keys[i].equals(key)) {
                return stripQuotes(values[i]);
            }
        }
        return "";
    }

    String[] getArray(String key) {
        for (int i = 0; i < size; i++) {
            if (keys[i].equals(key)) {
                var raw = values[i].trim();
                if (raw.startsWith("[")) raw = raw.substring(1);
                if (raw.endsWith("]")) raw = raw.substring(0, raw.length() - 1);
                raw = raw.trim();
                if (raw.isEmpty()) return new String[0];

                var parts = raw.split(",");
                var result = new String[parts.length];
                int count = 0;
                for (int j = 0; j < parts.length; j++) {
                    var item = parts[j].trim();
                    item = stripQuotes(item);
                    if (!item.isEmpty()) {
                        result[count] = item;
                        count++;
                    }
                }
                var trimmed = new String[count];
                for (int j = 0; j < count; j++) trimmed[j] = result[j];
                return trimmed;
            }
        }
        return new String[0];
    }

    String[] profileKeys() {
        var result = new String[profileCount];
        for (int i = 0; i < profileCount; i++) {
            result[i] = profiles[i];
        }
        return result;
    }

    private static String stripComment(String line) {
        boolean inSingle = false;
        boolean inDouble = false;
        for (int i = 0; i < line.length(); i++) {
            char ch = line.charAt(i);
            if (ch == '\'' && !inDouble) inSingle = !inSingle;
            else if (ch == '"' && !inSingle) inDouble = !inDouble;
            else if (ch == '#' && !inSingle && !inDouble) return line.substring(0, i);
        }
        return line;
    }

    private static String stripQuotes(String s) {
        s = s.trim();
        if ((s.startsWith("\"") && s.endsWith("\"")) || (s.startsWith("'") && s.endsWith("'"))) {
            return s.substring(1, s.length() - 1);
        }
        return s;
    }
}

File: rust/main/features/config/src/core/toml_config.rs
use std::collections::HashMap;
use std::path::Path;

use crate::api::ConfigProvider;

/// Pure-Rust TOML config parser. Supports single-line key=value and inline arrays.
pub struct DefaultConfigProvider {
    data: HashMap<String, String>,
    profiles: Vec<String>,
}

impl DefaultConfigProvider {
    pub fn load(path: &Path) -> Result<Self, String> {
        let content = std::fs::read_to_string(path)
            .map_err(|e| format!("Config file not found: {}: {e}", path.display()))?;

        let mut data = HashMap::new();
        let mut profiles = Vec::new();
        let mut current_section = String::new();

        for raw_line in content.lines() {
            let line = strip_comment(raw_line).trim().to_string();
            if line.is_empty() {
                continue;
            }

            if line.starts_with('[') && line.ends_with(']') {
                current_section = line[1..line.len() - 1].to_string();
                if let Some(profile_key) = current_section.strip_prefix("profiles.") {
                    profiles.push(profile_key.to_string());
                }
                continue;
            }

            if let Some(eq_pos) = line.find('=') {
                let key = line[..eq_pos].trim().to_string();
                let value = line[eq_pos + 1..].trim().to_string();
                let full_key = if current_section.is_empty() {
                    key
                } else {
                    format!("{}.{}", current_section, key)
                };
                data.insert(full_key, value);
            }
        }

        Ok(Self { data, profiles })
    }
}

impl ConfigProvider for DefaultConfigProvider {
    fn get_value(&self, key: &str) -> String {
        match self.data.get(key) {
            Some(raw) => strip_quotes(raw),
            None => String::new(),
        }
    }

    fn get_array(&self, key: &str) -> Vec<String> {
        let raw = match self.data.get(key) {
            Some(r) => r.clone(),
            None => return Vec::new(),
        };

        let raw = raw.trim_start_matches('[').trim_end_matches(']');
        if raw.trim().is_empty() {
            return Vec::new();
        }

        raw.split(',')
            .map(|item| item.trim().trim_matches('"').trim_matches('\'').to_string())
            .filter(|s| !s.is_empty())
            .collect()
    }

    fn profile_keys(&self) -> &[String] {
        &self.profiles
    }
}

/// Strip inline comments, respecting quoted strings.
fn strip_comment(line: &str) -> &str {
    let mut in_single = false;
    let mut in_double = false;

    for (i, ch) in line.char_indices() {
        match ch {
            '\'' if !in_double => in_single = !in_single,
            '"' if !in_single => in_double = !in_double,
            '#' if !in_single && !in_double => return &line[..i],
            _ => {}
        }
    }
    line
}

fn strip_quotes(s: &str) -> String {
    let s = s.trim();
    if (s.starts_with('"') && s.ends_with('"')) || (s.starts_with('\'') && s.ends_with('\'')) {
        s[1..s.len() - 1].to_string()
    } else {
        s.to_string()
    }
}
